Make XMonad.Util.NamedScratchpad more flexible

elkowar opened the pull request:
### Description

This PR adds more customizability to how scratchpads are initially started in `XMonad.Util.NamedScratchpad`, by adding two new exported functions:
- `customRunNamedScratchpadAction`
  (provides the option to customize the `X ()` action the scratchpad is launched by) 
- `spawnHereNamedScratchpadAction`
  (uses `XMonad.Actions.SpawnOn.spawnHere` to initially start the scratchpad on the workspace it was launched on)


### Checklist

  - [x] I've read [CONTRIBUTING.md](https://github.com/xmonad/xmonad/blob/master/CONTRIBUTING.md)

  - [x] I tested my changes with [xmonad-testing](https://github.com/xmonad/xmonad-testing)

  - [x] I updated the `CHANGES.md` file

byorgey commented on XMonad/Util/NamedScratchpad.hs:
```diff
@@ -115,30 +118,48 @@ findByName c s = listToMaybe $ filter ((s==) . name) c
+                               -> NamedScratchpads           -- ^ Named scratchpads configuration
+                               -> String                     -- ^ Scratchpad name
+                               -> X ()
+customRunNamedScratchpadAction = someNamedScratchpadAction (\f ws -> f $ head ws)
```
How do we know this use of `head` is safe?  Can this code be rewritten without using it, in a way that is more "obviously" safe?

byorgey commented:
This looks good to me, other than one dodgy-looking use of `head` that I'm wondering about.

elkowar commented:
I've refactored the module to now use `Data.NonEmpty`, which makes it obvious that this is now, in fact, safe. I've also refactored the `someNamedScratchpadAction` a bit in the process, to (in my opinion) improve the readability quite noticably. 
If you disagree with the code-style choices I've made here, i could obviously revert this to be closer to the original style.

I think making variable names a bit less abbreviated is a good idea for modules like these, because these are the modules that people like me start looking in to as examples on how to implement their own XMonad modules and features, so having the names be more obvious to people not that deep into the XMonad data structures makes it easier to start implementing custom functionality.

byorgey commented:
Looks great to me, thanks!  Sorry it's taken me a while to get around to it again.
